File: Managed/NextTurn.UE.Runtime/Core/Paths.cs
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

using NextTurn.UE;
using NextTurn.UE.Annotations;

namespace Unreal
{
    public static class Paths
    {
        public static unsafe string EngineDirectory
        {
            get
            {
                ScriptArray nativeResult;
                NativeMethods.GetEngineDirectory(&nativeResult);
                return StringMarshaler.ToManagedFinally(&nativeResult);
            }
        }

        public static unsafe string EngineConfigDirectory
        {
            get
            {
                ScriptArray nativeResult;
                NativeMethods.GetEngineConfigDirectory(&nativeResult);
                return StringMarshaler.ToManagedFinally(&nativeResult);
            }
        }

        public static unsafe string EngineContentDirectory
        {
            get
            {
                ScriptArray nativeResult;
                NativeMethods.GetEngineContentDirectory(&nativeResult);
                return StringMarshaler.ToManagedFinally(&nativeResult);
            }
        }

        public static unsafe string EnginePluginsDirectory
        {
            get
            {
                ScriptArray nativeResult;
                NativeMethods.GetEnginePluginsDirectory(&nativeResult);
                return StringMarshaler.ToManagedFinally(&nativeResult);
            }
        }

        public static unsafe string EngineSavedDirectory
        {
            get
            {
                ScriptArray nativeResult;
                NativeMethods.GetEngineSavedDirectory(&nativeResult);
                return StringMarshaler.ToManagedFinally(&nativeResult);
            }
        }

        public static unsafe string ProjectDirectory
        {
            get
            {
                ScriptArray nativeResult;
                NativeMethods.GetProjectDirectory(&nativeResult);
                return StringMarshaler.ToManagedFinally(&nativeResult);
            }
        }

        public static unsafe string ProjectConfigDirectory
        {
            get
            {
                ScriptArray nativeResult;
                NativeMethods.GetProjectConfigDirectory(&nativeResult);
                return StringMarshaler.ToManagedFinally(&nativeResult);
            }
        }

        public static unsafe string ProjectContentDirectory
        {
            get
            {
                ScriptArray nativeResult;
                NativeMethods.GetProjectContentDirectory(&nativeResult);
                return StringMarshaler.ToManagedFinally(&nativeResult);
            }
        }

        public static unsafe string ProjectPluginsDirectory
        {
            get
            {
                ScriptArray nativeResult;
                NativeMethods.GetProjectPluginsDirectory(&nativeResult);
                return StringMarshaler.ToManagedFinally(&nativeResult);
            }
        }

        public static unsafe string ProjectSavedDirectory
        {
            get
            {
                ScriptArray nativeResult;
                NativeMethods.GetProjectSavedDirectory(&nativeResult);
                return StringMarshaler.ToManagedFinally(&nativeResult);
            }
        }

        private static class NativeMethods
        {
            [Calli]
            public static extern unsafe void GetEngineDirectory(ScriptArray* nativeResult);

            [Calli]
            public static extern unsafe void GetEngineConfigDirectory(ScriptArray* nativeResult);

            [Calli]
            public static extern unsafe void GetEngineContentDirectory(ScriptArray* nativeResult);

            [Calli]
            public static extern unsafe void GetEnginePluginsDirectory(ScriptArray* nativeResult);

            [Calli]
            public static extern unsafe void GetEngineSavedDirectory(ScriptArray* nativeResult);

            [Calli]
            public static extern unsafe void GetProjectDirectory(ScriptArray* nativeResult);

            [Calli]
            public static extern unsafe void GetProjectConfigDirectory(ScriptArray* nativeResult);

            [Calli]
            public static extern unsafe void GetProjectContentDirectory(ScriptArray* nativeResult);

            [Calli]
            public static extern unsafe void GetProjectPluginsDirectory(ScriptArray* nativeResult);

            [Calli]
            public static extern unsafe void GetProjectSavedDirectory(ScriptArray* nativeResult);
        }
    }
}

File: Source/NextTurnRuntime/Private/Core/Paths.cpp
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

#include "NextTurnRuntime/Private/Exports.hpp"

#include "NextTurnRuntime/Private/MarshaledTypes.hpp"
#include "NextTurnRuntime/Public/Compiler.hpp"

#include "Runtime/Core/Public/CoreMinimal.h"
#include "Runtime/Core/Public/Misc/Paths.h"

ENABLE_WARNINGS;

#define TYPE_EXPORT_MANAGED_TYPE_NAME Paths

TYPE_EXPORT_START

void EXPORT_CALL_CONV GetEngineDirectory(FMarshaledString& OutResult)
{
	OutResult = FPaths::EngineDir();
}

void EXPORT_CALL_CONV GetEngineConfigDirectory(FMarshaledString& OutResult)
{
	OutResult = FPaths::EngineConfigDir();
}

void EXPORT_CALL_CONV GetEngineContentDirectory(FMarshaledString& OutResult)
{
	OutResult = FPaths::EngineContentDir();
}

void EXPORT_CALL_CONV GetEnginePluginsDirectory(FMarshaledString& OutResult)
{
	OutResult = FPaths::EnginePluginsDir();
}

void EXPORT_CALL_CONV GetEngineSavedDirectory(FMarshaledString& OutResult)
{
	OutResult = FPaths::EngineSavedDir();
}

void EXPORT_CALL_CONV GetProjectDirectory(FMarshaledString& OutResult)
{
	OutResult = FPaths::ProjectDir();
}

void EXPORT_CALL_CONV GetProjectConfigDirectory(FMarshaledString& OutResult)
{
	OutResult = FPaths::ProjectConfigDir();
}

void EXPORT_CALL_CONV GetProjectContentDirectory(FMarshaledString& OutResult)
{
	OutResult = FPaths::ProjectContentDir();
}

void EXPORT_CALL_CONV GetProjectPluginsDirectory(FMarshaledString& OutResult)
{
	OutResult = FPaths::ProjectPluginsDir();
}

void EXPORT_CALL_CONV GetProjectSavedDirectory(FMarshaledString& OutResult)
{
	OutResult = FPaths::ProjectSavedDir();
}

const FMemberSymbol Members[] =
{
	EXPORT_METHOD(GetEngineDirectory),
	EXPORT_METHOD(GetEngineConfigDirectory),
	EXPORT_METHOD(GetEngineContentDirectory),
	EXPORT_METHOD(GetEnginePluginsDirectory),
	EXPORT_METHOD(GetEngineSavedDirectory),
	EXPORT_METHOD(GetProjectDirectory),
	EXPORT_METHOD(GetProjectConfigDirectory),
	EXPORT_METHOD(GetProjectContentDirectory),
	EXPORT_METHOD(GetProjectPluginsDirectory),
	EXPORT_METHOD(GetProjectSavedDirectory),
};

TYPE_EXPORT_END

#undef TYPE_EXPORT_MANAGED_TYPE_NAME

RESTORE_WARNINGS;
